domain/src/main/java/org/oppia/android/domain/hintsandsolution/HintHandlerDebugImpl.kt

```diff
@@ -0,0 +1,126 @@
+package org.oppia.android.domain.hintsandsolution
+
+import org.oppia.android.app.model.HelpIndex
+import org.oppia.android.app.model.State
+import org.oppia.android.domain.devoptions.ShowAllHintsAndSolutionHelper
+import java.util.concurrent.locks.ReentrantLock
+import javax.inject.Inject
+import kotlin.concurrent.withLock
+
+/**
+ * Debug implementation of [HintHandler] that overrides the default mechanism of hints & solutions
+ * to show all hints and solutions based on availability, bypassing the wrong answer check and
+ * scheduling of hints, if 'Show all hints and solution' functionality is enabled in the
+ * 'Developer Options Menu'. If this functionality is disabled then it will fall back to
+ * [HintHandlerProdImpl] and provide its functionalities.
+ */
+class HintHandlerDebugImpl private constructor(
+  private val hintHandlerProdImpl: HintHandlerProdImpl,
+  private val showAllHintsAndSolutionHelper: ShowAllHintsAndSolutionHelper,
+  private val hintMonitor: HintHandler.HintMonitor
+) : HintHandler {
+
+  private val handlerLock = ReentrantLock()
+
+  override fun startWatchingForHintsInNewState(state: State) {
+    if (!showAllHintsAndSolutionHelper.getShowAllHintsAndSolution()) {
+      hintHandlerProdImpl.startWatchingForHintsInNewState(state)
+    } else {
+      handlerLock.withLock {
+        hintHandlerProdImpl.pendingState = state
+        hintMonitor.onHelpIndexChanged()
+        showAllHintsAndSolution()
+      }
+    }
+  }
+
+  override fun finishState(newState: State) {
+    if (!showAllHintsAndSolutionHelper.getShowAllHintsAndSolution()) {
+      hintHandlerProdImpl.finishState(newState)
+    } else {
+      handlerLock.withLock {
+        hintHandlerProdImpl.reset()
+        startWatchingForHintsInNewState(newState)
+      }
+    }
+  }
+
+  override fun handleWrongAnswerSubmission(wrongAnswerCount: Int) {
+    if (!showAllHintsAndSolutionHelper.getShowAllHintsAndSolution()) {
+      hintHandlerProdImpl.handleWrongAnswerSubmission(wrongAnswerCount)
+    }
+  }
+
+  override fun viewHint(hintIndex: Int) {
+    if (!showAllHintsAndSolutionHelper.getShowAllHintsAndSolution()) {
+      hintHandlerProdImpl.viewHint(hintIndex)
+    } else {
```

Comment:
Isn't this case invalid? I thought we were by default jumping straight to the EVERYTHING_REVEALED state?

Reply:
Following your suggestion to return EVERYTHING_REVEALED state, I have removed this case. Thanks for the suggestion.